core/src/test/java/org/apache/accumulo/core/file/rfile/MultiThreadedRFileTest.java

```diff
@@ -0,0 +1,387 @@
+    }
+  }
+
+  static Key nk(String row, String cf, String cq, String cv, long ts) {
+    return new Key(row.getBytes(), cf.getBytes(), cq.getBytes(), cv.getBytes(), ts);
+  }
+
+  static Value nv(String val) {
+    return new Value(val.getBytes());
+  }
+
+  public AccumuloConfiguration conf = null;
+
+  @Test
+  public void testMultipleReaders() throws IOException {
+    final List<Throwable> threadExceptions = Collections.synchronizedList(new ArrayList<Throwable>());
+    Map<String,MutableInt> messages = new HashMap<String,MutableInt>();
+    Map<String,String> stackTrace = new HashMap<String,String>();
+
+    final TestRFile trfBase = new TestRFile(conf);
+
+    writeData(trfBase);
+
+    trfBase.openReader();
+
+    try {
+
+      validate(trfBase);
+
+      final TestRFile trfBaseCopy = trfBase.deepCopy();
+
+      validate(trfBaseCopy);
+
+      // now start up multiple RFile deepcopies
+      int maxThreads = 10;
+      String name = "MultiThreadedRFileTestThread";
+      ThreadPoolExecutor pool = new ThreadPoolExecutor(maxThreads + 1, maxThreads + 1, 5 * 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
+          new NamingThreadFactory(name));
+      pool.allowCoreThreadTimeOut(true);
+      try {
+        Runnable runnable = new Runnable() {
+          @Override
+          public void run() {
+            try {
+              TestRFile trf = trfBase;
+              synchronized (trfBaseCopy) {
+                trf = trfBaseCopy.deepCopy();
+              }
+              validate(trf);
+            } catch (Throwable t) {
+              threadExceptions.add(t);
+            }
+          }
+        };
+        for (int i = 0; i < maxThreads; i++) {
+          pool.submit(runnable);
+        }
+      } finally {
+        pool.shutdown();
+        try {
+          pool.awaitTermination(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+      }
+
+      for (Throwable t : threadExceptions) {
+        String msg = t.getClass() + " : " + t.getMessage();
+        if (!messages.containsKey(msg)) {
+          messages.put(msg, new MutableInt(1));
+        } else {
+          messages.get(msg).increment();
+        }
+        StringWriter string = new StringWriter();
+        PrintWriter writer = new PrintWriter(string);
+        t.printStackTrace(writer);
+        writer.flush();
+        stackTrace.put(msg, string.getBuffer().toString());
+      }
+    } finally {
+      trfBase.closeReader();
+      trfBase.close();
+    }
+
+    for (String message : messages.keySet()) {
+      System.out.println(messages.get(message) + ": " + message);
+      System.out.println(stackTrace.get(message));
+    }
+
+    assertTrue(threadExceptions.isEmpty());
+  }
+
+  private void validate(TestRFile trf) throws IOException {
+    Random random = new Random();
+    for (int iteration = 0; iteration < 10; iteration++) {
+      int part = random.nextInt(4);
```

Comment:
Note that caching seems to have obfuscated the errors that we were seeing, so no data or index cache is supplied in this case.